### PR TITLE
[WIP] MultiAgent A2C

### PR DESCRIPTION
**Pull request by AdityaKapoor74**
Multi Agent Integration

### CODE REVIEW

File: genrl/marl/maa2c/a2c.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+#         self.policy = nn.Linear(256, self.action_dim)
+#         # torch.nn.init.kaiming_normal_(self.policy.weight, mode='fan_in')
+#         torch.nn.init.xavier_uniform_(self.policy.weight)
+
+#     def forward(self, x):
+#         x_s = F.relu(self.shared_layer(x))
+#         qval = self.value(x_s)
+#         policy = self.policy(x_s)
+
+#         return policy,qval
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c.py:31]: Can you delete **all** the comments?

File: genrl/marl/maa2c/a2c.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        self.value = nn.Linear(256, 1)
+        # torch.nn.init.kaiming_normal_(self.value.weight, mode='fan_in')
+        torch.nn.init.xavier_uniform_(self.value.weight)
+        # torch.nn.init.xavier_normal_(self.value.weight)
+
+        self.policy = nn.Linear(256, self.action_dim)
+        # torch.nn.init.kaiming_normal_(self.policy.weight, mode='fan_in')
+        torch.nn.init.xavier_uniform_(self.policy.weight)
+        # torch.nn.init.xavier_normal_(self.policy.weight)
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c.py:60]: Although its ok for now, we need to define a way to create these shared layer Neural Networks easily in a single call. 

cc: @ajaysub110

File: genrl/marl/maa2c/a2c_agent.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    
+    self.num_agents = self.env.n
+
+    self.input_dim = env.observation_space[0].shape[0]
+    self.action_dim = self.env.action_space[0].n
+    self.actorcritic = CentralizedActorCritic(self.input_dim,self.action_dim).to(self.device)
+    model_path = ".tests/models/four_agents/actorcritic_network_no_comms_discounted_rewards_lr_2e-4_with_grad_norm_0.5_entropy_pen_0.008_xavier_uniform_init_clamp_logs.pt"
+    self.actorcritic.load_state_dict(torch.load(model_path,map_location=torch.device('cpu')))
+
+    self.MSELoss = nn.MSELoss()
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:29]: Not required? Can directly use F.mse_loss right where the loss computation is happening?

File: genrl/marl/maa2c/a2c_agent.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    
+    self.num_agents = self.env.n
+
+    self.input_dim = env.observation_space[0].shape[0]
+    self.action_dim = self.env.action_space[0].n
+    self.actorcritic = CentralizedActorCritic(self.input_dim,self.action_dim).to(self.device)
+    model_path = ".tests/models/four_agents/actorcritic_network_no_comms_discounted_rewards_lr_2e-4_with_grad_norm_0.5_entropy_pen_0.008_xavier_uniform_init_clamp_logs.pt"
+    self.actorcritic.load_state_dict(torch.load(model_path,map_location=torch.device('cpu')))
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:27]: Model loading should ideally happen in a separate method such as `setup_model` like in the other Deep agents.

File: genrl/marl/maa2c/a2c_agent.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    self.MSELoss = nn.MSELoss()
+    self.actorcritic_optimizer = optim.Adam(self.actorcritic.parameters(),lr=lr)
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:30]: Could you rename this to just `self.optimizer` if there is no other optimizer here?
Also, should be shifted to a separate method like `setup_model`

File: genrl/marl/maa2c/a2c_agent.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    del state
+    dist = F.softmax(logits,dim=0)
+    # print("dist:",dist)
+    del logits
+    # print('dist: ', dist)
+    probs = Categorical(dist)
+    # print("PROBS:",probs)
+    index = probs.sample().cpu().detach().item()
+    
+    return index
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:45]: This can be shifted to 'centralised_actor_critic` class if `get_action` will look similar in all algorithms.

File: genrl/marl/maa2c/a2c_agent.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+
+  def update(self,global_state_batch,global_next_state_batch,global_actions_batch,rewards):
+    
+
+    curr_logits,curr_Q = self.actorcritic.forward(global_state_batch)
+    # _,next_Q = self.actorcritic.forward(global_next_state_batch)
+    # estimated_Q = rewards.unsqueeze(dim=2) + self.gamma*next_Q
+    # print("REWARDS",rewards.shape)
+    discounted_rewards = np.asarray([[torch.sum(torch.FloatTensor([self.gamma**i for i in range(rewards[k][j:].size(0))])* rewards[k][j:]) for j in range(rewards.size(0))] for k in range(self.num_agents)])
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:55]: Although much faster, list comprehension losses brevity. Do you think we should separate line 55 as a separate function and do a proper for loop?

File: genrl/marl/maa2c/a2c_agent.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    # print("ESTIMATES Q")
+    # print(estimated_Q)
+    # print(estimated_Q.shape)
+    # print("CURRENT LOGITS")
+    # print(curr_logits)
+    # print(curr_logits.shape)
+
+    
+
+    # critic_loss = self.MSELoss(curr_Q,estimated_Q.detach())
```
Comment (Sharad24) [inline genrl/marl/maa2c/a2c_agent.py:76]: So many comments
😅

File: genrl/marl/maa2c/maa2c.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    actions = torch.LongTensor([sars[2] for sars in trajectory]).to(self.device)
+    rewards = torch.FloatTensor([sars[3] for sars in trajectory]).to(self.device)
+
+    critic_loss,policy_loss,entropy,total_loss,grad_norm = self.agents.update(states,next_states,actions,rewards)
+
+    self.writer.add_scalar('Loss/Entropy loss',entropy,episode)
+    self.writer.add_scalar('Loss/Value Loss',critic_loss,episode)
+    self.writer.add_scalar('Loss/Policy Loss',policy_loss,episode)
+    self.writer.add_scalar('Loss/Total Loss',total_loss,episode)
+    self.writer.add_scalar('Gradient Normalization/Grad Norm',grad_norm,episode)
```
Comment (Sharad24) [inline genrl/marl/maa2c/maa2c.py:46]: I think what you could do is from the `update` method return a dictionary and directly pass that to the logger and these statements would be reduced to 1. 

What do you think?

File: genrl/marl/maa2c/maa2c.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+        # print(rewards)
+
+
+        episode_reward += np.sum(rewards)
+
+        # print(dones)
+        # print(all(dones))
+
+        if all(dones) or step == max_steps-1:
```
Comment (Sharad24) [inline genrl/marl/maa2c/maa2c.py:68]: Is there any way to use the On Policy trainer? Its very modular and you have to define your agent in 3 calls. `collect_rollouts`, `get_traj_loss` and `update_policy`.

I think till now what I've seen it seems it can fit well. What do you think?

File: genrl/marl/maa2c/maa2c.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+          dones = [0 for _ in range(self.num_agents)]
+          trajectory.append([states,next_states,actions,rewards])
+          states = next_states
+      
+#       make a directory called models
+      if episode%500:
+        torch.save(self.agents.actorcritic.state_dict(), ".tests//models/four_agents/actorcritic_network_no_comms_discounted_rewards_lr_2e-4_with_grad_norm_0.5_entropy_pen_0.008_xavier_uniform_init_clamp_logs.pt")
+      
+        
+      self.update(trajectory,episode)
```
Comment (Sharad24) [inline genrl/marl/maa2c/maa2c.py:90]: You could get rid of a lot of code here if we use the On Policy Trainer.

File: genrl/marl/maa2c/simple_spread_test.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    # create world
+    world = scenario.make_world()
+    # create multiagent environment
+    if benchmark:
+        env = MultiAgentEnv(world, scenario.reset_world, scenario.reward, scenario.observation, scenario.benchmark_data, scenario.isFinished)
+    else:
+        env = MultiAgentEnv(world, scenario.reset_world, scenario.reward, scenario.observation, None, scenario.isFinished)
+    return env
+
+
```
Comment (Sharad24) [inline genrl/marl/maa2c/simple_spread_test.py:20]: Let's shift this to the `examples` directory?

**Comment by Sharad24** [comment]
Also, do you think we should rename `marl` to `multiagent` or even `multi`?

**Comment by Sharad24** [comment]
Some other notes:
1. If you have any questions about the OnPolicyTrainer or Logger, please feel free to ask as we don't have good examples right now. But it's great to find that we need to add something on this.  
2. Could you configure pre-commit? Its pretty easy, just install `pip install pre-commit` followed by `pre-commit install`. You may have to run git commit a few times after this, to create a single commit as pre-commit is doing its job.

File: genrl/marl/maa2c/simple_spread_test.py (hunk 12 above)
Comment (Sharad24) [reply genrl/marl/maa2c/simple_spread_test.py:20]: Nvm, I got confused that this is an example

**Comment by Sharad24** [comment]
@AdityaKapoor74 Let me know if you are stuck on anything here, happy to help. Ideally we should merge this before MADDPG. Can probably open another PR for that.

**Comment by AdityaKapoor74** [comment]
> @AdityaKapoor74 Let me know if you are stuck on anything here, happy to help. Ideally we should merge this before MADDPG. Can probably open another PR for that.

I have got a little busy lately. Hopefully by this week I will have this PR sorted out and I can raise a PR for MADDPG as well @Sharad24

**Comment by Sharad24** [comment]
Ok

File: genrl/marl/maa2c/a2c_agent.py (hunk 8 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/a2c_agent.py:76]: Yeah I was just making sure that the tensors shapes don't mismatch :package:

File: genrl/marl/maa2c/a2c_agent.py (hunk 5 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/a2c_agent.py:30]: Cool!

File: genrl/marl/maa2c/maa2c.py (hunk 9 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/maa2c.py:46]: Yeah, I just had a look at the Tensorboard logger. It can be done

File: genrl/marl/maa2c/maa2c.py (hunk 10 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/maa2c.py:68]: It can be done. I am not sure how the environment will be loaded though. Could you explain?

File: genrl/marl/maa2c/maa2c.py (hunk 11 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/maa2c.py:90]: Yeah. Can be done. I am not sure about two things in the OnPolicyTrainer. What do the variables run_num and evaluate_episodes do?

File: genrl/marl/maa2c/a2c_agent.py (hunk 6 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/a2c_agent.py:45]: It is environment dependent but yeah it can done.

File: genrl/marl/maa2c/a2c_agent.py (hunk 7 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/a2c_agent.py:55]: We can. Actually I was thinking of a much better way to get the Value_target.. Maybe we can discuss?

File: genrl/marl/maa2c/a2c_agent.py (hunk 3 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/a2c_agent.py:29]: Yeah. Actually it is better to use Huber loss.

File: genrl/marl/maa2c/maa2c.py (hunk 11 above)
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/maa2c.py:90]: > Yeah. Can be done. I am not sure about two things in the OnPolicyTrainer. What do the variables run_num and evaluate_episodes do?

Never mind. I got it

File: genrl/marl/maa2c/logger.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+    def write(self, kvs: Dict[str, Any]) -> None:
+        """
+        Add entry to logger
+
+        :param kvs: Entries to be logged
+        :type kvs: dict
+        """
+        for key, value in kvs.items():
+            self.writer.add_scalar(key, value, kvs["Episode"])
```
[the remark this answers is not in the document]
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/logger.py:29]: Done already :D

File: genrl/marl/maa2c/multiagent/rendering.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+"""
+from __future__ import division
+import os
+import six
+import sys
+
+if "Apple" in sys.version:
+    if 'DYLD_FALLBACK_LIBRARY_PATH' in os.environ:
+        os.environ['DYLD_FALLBACK_LIBRARY_PATH'] += ':/usr/lib'
+        # (JDS 2016/04/15): avoid bug on Anaconda 2.3.0 / Yosemite
```
[the remark this answers is not in the document]
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/multiagent/rendering.py:12]: That's done by OpenAI

File: genrl/marl/maa2c/multiagent/rendering.py [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+
+try:
+    import pyglet
+except ImportError as e:
+    reraise(suffix="HINT: you can install pyglet directly via 'pip install pyglet'. But if you really just want to install all Gym dependencies and not have to think about it, 'pip install -e .[all]' or 'pip install gym[all]' will do it.")
+
+try:
+    from pyglet.gl import *
+except ImportError as e:
+    reraise(prefix="Error occured while running `from pyglet.gl import *`",suffix="HINT: make sure you have OpenGL install. On Ubuntu, you can run 'apt-get install python-opengl'. If you're running on a server, you may need a virtual frame buffer; something like this should work: 'xvfb-run -s \"-screen 0 1400x900x24\" python <your_script.py>'")
```
[the remark this answers is not in the document]
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/multiagent/rendering.py:25]: Yeah sure

File: genrl/marl/maa2c/multiagent/rendering.py [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+
+import math
+import numpy as np
+
+RAD2DEG = 57.29577951308232
```
[the remark this answers is not in the document]
Comment (AdityaKapoor74) [reply genrl/marl/maa2c/multiagent/rendering.py:30]: Again its the gym environment, which I didn't really change

**Comment by AdityaKapoor74** [comment]
@Sharad24 let me know about the MAA2C code. We can change the environment setting once that's taken care of. Also, there are multiagent environments by Unity. I haven't checked them out. If they are similar to OpenAI then we can have a common wrapper for both

**Comment by Sharad24** [comment]
`__init__.py` files have nothing in them, can you check the codebase and see which classes we should be exposing to the user?

**Comment by Sharad24** [comment]
Have you installed `pre-commit`? Please run linting, etc through that? Let me know if you need help with that. 

`pip install pre-commit`
`pre-commit install`

File: genrl/marl/maa2c/generate_gif.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+	# init a new episode
+	obs = env.reset()
+	# init the img var with the starting state of the env
+	img = env.render(mode='rgb_array')[0]
+
+	for i in range(n_images):
+	  # At each step, append an image to list
+	  images.append(img)
+
```
Comment (Sharad24) [inline genrl/marl/maa2c/generate_gif.py:30]: This is only an example? Please move it to `examples`
